File: adapters/harbor/bundles/rogue_singleplayer_gold/reference/gold_python/source_chase.py
"""Source-faithful Rogue chase movement decision slices."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from source_rogue import NUMCOLS, NUMLINES, SCROLL, RogueRng, step_ok


ISHUH = 0o001000
ISHASTE = 0o000100
ISSLOW = 0o100000
S_SCARE = 10


@dataclass
class Coord:
    y: int
    x: int

    def to_dict(self) -> dict[str, int]:
        return {"y": self.y, "x": self.x}


@dataclass
class ChaseThing:
    monster_type: str
    pos: Coord
    flags: int
    turn: bool = True
    disguise: str | None = None

    def __post_init__(self) -> None:
        if self.disguise is None:
            self.disguise = self.monster_type

    def to_dict(self) -> dict[str, Any]:
        return {
            "type": self.monster_type,
            "pos": self.pos.to_dict(),
            "flags": self.flags,
            "turn": self.turn,
            "disguise": self.disguise,
        }


@dataclass
class ChaseObject:
    obj_type: str
    which: int
    pos: Coord


@dataclass
class ChaseMap:
    terrain: dict[tuple[int, int], str] = field(default_factory=dict)
    objects: list[ChaseObject] = field(default_factory=list)
    monsters: list[ChaseThing] = field(default_factory=list)

    def chat(self, coord: Coord) -> str:
        return self.terrain.get((coord.y, coord.x), ".")

    def moat(self, coord: Coord) -> ChaseThing | None:
        for monster in self.monsters:
            if coord_eq(monster.pos, coord):
                return monster
        return None

    def winat(self, coord: Coord) -> str:
        monster = self.moat(coord)
        if monster is not None:
            return monster.disguise or monster.monster_type
        return self.chat(coord)

    def object_at(self, coord: Coord) -> ChaseObject | None:
        for obj in self.objects:
            if coord_eq(obj.pos, coord):
                return obj
        return None


def chase(rng: RogueRng, game_map: ChaseMap, thing: ChaseThing, target: Coord, hero: Coord) -> dict[str, Any]:
    trace: dict[str, Any] = {"branch": "direct", "candidates": []}
    random_branch = False
    if thing.flags & ISHUH:
        confused_roll = rng.rnd(5)
        trace["confused_roll"] = confused_roll
        random_branch = confused_roll != 0
    if not random_branch and thing.monster_type == "P":
        phantom_roll = rng.rnd(5)
        trace["phantom_roll"] = phantom_roll
        random_branch = phantom_roll == 0
    if not random_branch and thing.monster_type == "B":
        bat_roll = rng.rnd(2)
        trace["bat_roll"] = bat_roll
        random_branch = bat_roll == 0

    if random_branch:
        trace["branch"] = "random"
        chosen = rndmove(rng, thing)
        curdist = dist(chosen, target)
        clear_roll = rng.rnd(20)
        trace["clear_roll"] = clear_roll
        if clear_roll == 0:
            thing.flags &= ~ISHUH
    else:
        curdist = dist(thing.pos, target)
        chosen = Coord(thing.pos.y, thing.pos.x)
        plcnt = 1
        ey = thing.pos.y + 1
        if ey >= NUMLINES - 1:
            ey = NUMLINES - 2
        ex = thing.pos.x + 1
        if ex >= NUMCOLS:
            ex = NUMCOLS - 1
        for x in range(thing.pos.x - 1, ex + 1):
            if x < 0:
                continue
            for y in range(thing.pos.y - 1, ey + 1):
                tryp = Coord(y, x)
                candidate: dict[str, Any] = {"pos": tryp.to_dict()}
                if not diag_ok(game_map, thing.pos, tryp):
                    candidate["skip"] = "diag"
                    trace["candidates"].append(candidate)
                    continue
                ch = game_map.winat(tryp)
                candidate["ch"] = ch
                if not step_ok(ch):
                    candidate["skip"] = "blocked"
                    trace["candidates"].append(candidate)
                    continue
                obj = game_map.object_at(tryp)
                if ch == SCROLL and obj is not None and obj.which == S_SCARE:
                    candidate["skip"] = "scare_scroll"
                    trace["candidates"].append(candidate)
                    continue
                blocker = game_map.moat(tryp)
                if blocker is not None and blocker.monster_type == "X":
                    candidate["skip"] = "xeroc"
                    trace["candidates"].append(candidate)
                    continue
                thisdist = dist(tryp, target)
                candidate["dist"] = thisdist
                if thisdist < curdist:
                    plcnt = 1
                    chosen = Coord(tryp.y, tryp.x)
                    curdist = thisdist
                    candidate["chosen"] = True
                elif thisdist == curdist:
                    plcnt += 1
                    tie_roll = rng.rnd(plcnt)
                    candidate["tie_roll"] = tie_roll
                    candidate["plcnt"] = plcnt
                    if tie_roll == 0:
                        chosen = Coord(tryp.y, tryp.x)
                        curdist = thisdist
                        candidate["chosen"] = True
                trace["candidates"].append(candidate)

    keep_chasing = curdist != 0 and not coord_eq(chosen, hero)
    return {
        "chosen": chosen.to_dict(),
        "curdist": curdist,
        "keep_chasing": keep_chasing,
        "thing": thing.to_dict(),
        "rng_seed": rng.seed,
        "trace": trace,
    }


def move_monst_schedule(thing: ChaseThing, do_chase_results: list[int]) -> dict[str, Any]:
    calls = 0
    returned = 0
    if not (thing.flags & ISSLOW) or thing.turn:
        calls += 1
        if do_chase_results and do_chase_results[0] == -1:
            returned = -1
            return _move_schedule_payload(thing, calls, returned)
    if thing.flags & ISHASTE:
        calls += 1
        if len(do_chase_results) > 1 and do_chase_results[1] == -1:
            returned = -1
            return _move_schedule_payload(thing, calls, returned)
    thing.turn = not thing.turn
    return _move_schedule_payload(thing, calls, returned)


def source_chase_report() -> dict[str, Any]:
    return {
        "schema": "gamebench.rogue.source_chase.v1",
        "chase_cases": [_run_chase_case(case) for case in _chase_cases()],
        "move_monst": [_run_move_case(case) for case in _move_cases()],
    }


def _run_chase_case(case: dict[str, Any]) -> dict[str, Any]:
    game_map = ChaseMap()
    for tile in case.get("tiles", []):
        game_map.terrain[(tile["y"], tile["x"])] = tile["ch"]
    for obj in case.get("objects", []):
        game_map.objects.append(ChaseObject(obj["type"], obj["which"], Coord(obj["y"], obj["x"])))
    for monster in case.get("monsters", []):
        game_map.monsters.append(ChaseThing(monster["type"], Coord(monster["y"], monster["x"]), monster.get("flags", 0), True, monster.get("disguise")))
    thing = ChaseThing(case["type"], Coord(case["pos"][0], case["pos"][1]), case.get("flags", 0), case.get("turn", True), case.get("disguise"))
    target = Coord(case["target"][0], case["target"][1])
    hero = Coord(case["hero"][0], case["hero"][1])
    outcome = chase(RogueRng(case["seed"]), game_map, thing, target, hero)
    return {"name": case["name"], "seed": case["seed"], "outcome": outcome}


def _run_move_case(case: dict[str, Any]) -> dict[str, Any]:
    thing = ChaseThing(case["type"], Coord(5, 5), case["flags"], case["turn"])
    return {"name": case["name"], "outcome": move_monst_schedule(thing, case.get("results", []))}


def _chase_cases() -> list[dict[str, Any]]:
    return [
        {"name": "direct_east", "seed": 1, "type": "K", "pos": [5, 5], "target": [5, 7], "hero": [5, 7]},
        {"name": "tie_break", "seed": 7, "type": "K", "pos": [5, 5], "target": [7, 7], "hero": [9, 9]},
        {"name": "wall_blocks_diagonal", "seed": 7, "type": "K", "pos": [5, 5], "target": [4, 4], "hero": [4, 4], "tiles": [{"y": 4, "x": 5, "ch": "|"}, {"y": 5, "x": 4, "ch": "-"}]},
        {"name": "scare_scroll_skip", "seed": 7, "type": "K", "pos": [5, 5], "target": [5, 4], "hero": [5, 4], "tiles": [{"y": 5, "x": 4, "ch": SCROLL}], "objects": [{"type": SCROLL, "which": S_SCARE, "y": 5, "x": 4}]},
        {"name": "ordinary_scroll_allowed", "seed": 7, "type": "K", "pos": [5, 5], "target": [5, 4], "hero": [5, 4], "tiles": [{"y": 5, "x": 4, "ch": SCROLL}], "objects": [{"type": SCROLL, "which": 1, "y": 5, "x": 4}]},
        {"name": "xeroc_skip", "seed": 7, "type": "K", "pos": [5, 5], "target": [5, 4], "hero": [5, 4], "monsters": [{"type": "X", "y": 5, "x": 4, "disguise": ":"}]},
        {"name": "confused_random", "seed": 5, "type": "K", "flags": ISHUH, "pos": [5, 5], "target": [7, 7], "hero": [7, 7]},
        {"name": "phantom_random", "seed": 1, "type": "P", "pos": [5, 5], "target": [7, 7], "hero": [7, 7]},
        {"name": "bat_random", "seed": 1, "type": "B", "pos": [5, 5], "target": [7, 7], "hero": [7, 7]},
    ]


def _move_cases() -> list[dict[str, Any]]:
    return [
        {"name": "normal_one_chase", "type": "K", "flags": 0, "turn": True, "results": [0]},
        {"name": "slow_turn_skips", "type": "K", "flags": ISSLOW, "turn": False, "results": []},
        {"name": "slow_turn_moves", "type": "K", "flags": ISSLOW, "turn": True, "results": [0]},
        {"name": "haste_two_chases", "type": "K", "flags": ISHASTE, "turn": True, "results": [0, 0]},
        {"name": "first_chase_stops", "type": "K", "flags": ISHASTE, "turn": True, "results": [-1, 0]},
        {"name": "second_chase_stops", "type": "K", "flags": ISHASTE, "turn": True, "results": [0, -1]},
    ]


def rndmove(rng: RogueRng, thing: ChaseThing) -> Coord:
    y = thing.pos.y + rng.rnd(3) - 1
    x = thing.pos.x + rng.rnd(3) - 1
    return Coord(y, x)


def diag_ok(game_map: ChaseMap, start: Coord, end: Coord) -> bool:
    if end.x < 0 or end.x >= NUMCOLS or end.y <= 0 or end.y >= NUMLINES - 1:
        return False
    if end.x == start.x or end.y == start.y:
        return True
    return step_ok(game_map.chat(Coord(end.y, start.x))) and step_ok(game_map.chat(Coord(start.y, end.x)))


def dist(first: Coord, second: Coord) -> int:
    return (second.x - first.x) * (second.x - first.x) + (second.y - first.y) * (second.y - first.y)


def coord_eq(first: Coord, second: Coord) -> bool:
    return first.x == second.x and first.y == second.y


def _move_schedule_payload(thing: ChaseThing, calls: int, returned: int) -> dict[str, Any]:
    return {"calls": calls, "returned": returned, "thing": thing.to_dict()}

File: tasks/rogue-singleplayer/gold_rust/src/source_chase.rs
use crate::{step_ok, RogueRng, NUMCOLS, NUMLINES, SCROLL};
use serde_json::{json, Value};

const ISHUH: i32 = 0o001000;
const ISHASTE: i32 = 0o000100;
const ISSLOW: i32 = 0o100000;
const S_SCARE: i32 = 10;

#[derive(Clone, Copy)]
pub struct Coord {
    pub y: i32,
    pub x: i32,
}

#[derive(Clone)]
pub struct ChaseThing {
    pub monster_type: char,
    pub pos: Coord,
    pub flags: i32,
    pub turn: bool,
    pub disguise: char,
}

#[derive(Clone)]
pub struct ChaseObject {
    pub obj_type: char,
    pub which: i32,
    pub pos: Coord,
}

pub struct ChaseMap {
    pub terrain: Vec<(Coord, char)>,
    pub objects: Vec<ChaseObject>,
    pub monsters: Vec<ChaseThing>,
}

pub fn source_chase_report() -> Value {
    json!({
        "schema": "gamebench.rogue.source_chase.v1",
        "chase_cases": chase_cases().into_iter().map(run_chase_case).collect::<Vec<_>>(),
        "move_monst": move_cases().into_iter().map(run_move_case).collect::<Vec<_>>(),
    })
}

pub fn chase(
    rng: &mut RogueRng,
    game_map: &ChaseMap,
    thing: &mut ChaseThing,
    target: Coord,
    hero: Coord,
) -> Value {
    let mut trace = json!({"branch": "direct", "candidates": []});
    let mut random_branch = false;
    if thing.flags & ISHUH != 0 {
        let confused_roll = rng.rnd(5);
        trace
            .as_object_mut()
            .unwrap()
            .insert("confused_roll".to_string(), json!(confused_roll));
        random_branch = confused_roll != 0;
    }
    if !random_branch && thing.monster_type == 'P' {
        let phantom_roll = rng.rnd(5);
        trace
            .as_object_mut()
            .unwrap()
            .insert("phantom_roll".to_string(), json!(phantom_roll));
        random_branch = phantom_roll == 0;
    }
    if !random_branch && thing.monster_type == 'B' {
        let bat_roll = rng.rnd(2);
        trace
            .as_object_mut()
            .unwrap()
            .insert("bat_roll".to_string(), json!(bat_roll));
        random_branch = bat_roll == 0;
    }

    let chosen;
    let curdist;
    if random_branch {
        trace
            .as_object_mut()
            .unwrap()
            .insert("branch".to_string(), json!("random"));
        chosen = rndmove(rng, thing);
        curdist = dist(chosen, target);
        let clear_roll = rng.rnd(20);
        trace
            .as_object_mut()
            .unwrap()
            .insert("clear_roll".to_string(), json!(clear_roll));
        if clear_roll == 0 {
            thing.flags &= !ISHUH;
        }
    } else {
        let mut local_curdist = dist(thing.pos, target);
        let mut local_chosen = thing.pos;
        let mut plcnt = 1;
        let mut ey = thing.pos.y + 1;
        if ey >= NUMLINES as i32 - 1 {
            ey = NUMLINES as i32 - 2;
        }
        let mut ex = thing.pos.x + 1;
        if ex >= NUMCOLS as i32 {
            ex = NUMCOLS as i32 - 1;
        }
        for x in (thing.pos.x - 1)..=ex {
            if x < 0 {
                continue;
            }
            for y in (thing.pos.y - 1)..=ey {
                let tryp = Coord { y, x };
                let mut candidate = json!({"pos": coord_json(tryp)});
                if !diag_ok(game_map, thing.pos, tryp) {
                    candidate
                        .as_object_mut()
                        .unwrap()
                        .insert("skip".to_string(), json!("diag"));
                    push_candidate(&mut trace, candidate);
                    continue;
                }
                let ch = game_map.winat(tryp);
                candidate
                    .as_object_mut()
                    .unwrap()
                    .insert("ch".to_string(), json!(char_json(ch)));
                if !step_ok(ch) {
                    candidate
                        .as_object_mut()
                        .unwrap()
                        .insert("skip".to_string(), json!("blocked"));
                    push_candidate(&mut trace, candidate);
                    continue;
                }
                if ch == SCROLL {
                    if let Some(obj) = game_map.object_at(tryp) {
                        if obj.obj_type == SCROLL && obj.which == S_SCARE {
                            candidate
                                .as_object_mut()
                                .unwrap()
                                .insert("skip".to_string(), json!("scare_scroll"));
                            push_candidate(&mut trace, candidate);
                            continue;
                        }
                    }
                }
                if let Some(blocker) = game_map.moat(tryp) {
                    if blocker.monster_type == 'X' {
                        candidate
                            .as_object_mut()
                            .unwrap()
                            .insert("skip".to_string(), json!("xeroc"));
                        push_candidate(&mut trace, candidate);
                        continue;
                    }
                }
                let thisdist = dist(tryp, target);
                candidate
                    .as_object_mut()
                    .unwrap()
                    .insert("dist".to_string(), json!(thisdist));
                if thisdist < local_curdist {
                    plcnt = 1;
                    local_chosen = tryp;
                    local_curdist = thisdist;
                    candidate
                        .as_object_mut()
                        .unwrap()
                        .insert("chosen".to_string(), json!(true));
                } else if thisdist == local_curdist {
                    plcnt += 1;
                    let tie_roll = rng.rnd(plcnt);
                    candidate
                        .as_object_mut()
                        .unwrap()
                        .insert("tie_roll".to_string(), json!(tie_roll));
                    candidate
                        .as_object_mut()
                        .unwrap()
                        .insert("plcnt".to_string(), json!(plcnt));
                    if tie_roll == 0 {
                        local_chosen = tryp;
                        local_curdist = thisdist;
                        candidate
                            .as_object_mut()
                            .unwrap()
                            .insert("chosen".to_string(), json!(true));
                    }
                }
                push_candidate(&mut trace, candidate);
            }
        }
        chosen = local_chosen;
        curdist = local_curdist;
    }
    let keep_chasing = curdist != 0 && !coord_eq(chosen, hero);
    json!({
        "chosen": coord_json(chosen),
        "curdist": curdist,
        "keep_chasing": keep_chasing,
        "thing": thing_json(thing),
        "rng_seed": rng.seed,
        "trace": trace,
    })
}

pub fn move_monst_schedule(thing: &mut ChaseThing, do_chase_results: &[i32]) -> Value {
    let mut calls = 0;
    let mut returned = 0;
    if thing.flags & ISSLOW == 0 || thing.turn {
        calls += 1;
        if !do_chase_results.is_empty() && do_chase_results[0] == -1 {
            returned = -1;
            return move_schedule_payload(thing, calls, returned);
        }
    }
    if thing.flags & ISHASTE != 0 {
        calls += 1;
        if do_chase_results.len() > 1 && do_chase_results[1] == -1 {
            returned = -1;
            return move_schedule_payload(thing, calls, returned);
        }
    }
    thing.turn = !thing.turn;
    move_schedule_payload(thing, calls, returned)
}

#[derive(Clone)]
struct ChaseCase {
    name: &'static str,
    seed: i32,
    monster_type: char,
    flags: i32,
    pos: Coord,
    target: Coord,
    hero: Coord,
    tiles: Vec<(Coord, char)>,
    objects: Vec<ChaseObject>,
    monsters: Vec<ChaseThing>,
}

struct MoveCase {
    name: &'static str,
    monster_type: char,
    flags: i32,
    turn: bool,
    results: Vec<i32>,
}

fn run_chase_case(case: ChaseCase) -> Value {
    let game_map = ChaseMap {
        terrain: case.tiles,
        objects: case.objects,
        monsters: case.monsters,
    };
    let mut thing = ChaseThing {
        monster_type: case.monster_type,
        pos: case.pos,
        flags: case.flags,
        turn: true,
        disguise: case.monster_type,
    };
    let mut rng = RogueRng::new(case.seed);
    json!({"name": case.name, "seed": case.seed, "outcome": chase(&mut rng, &game_map, &mut thing, case.target, case.hero)})
}

fn run_move_case(case: MoveCase) -> Value {
    let mut thing = ChaseThing {
        monster_type: case.monster_type,
        pos: coord(5, 5),
        flags: case.flags,
        turn: case.turn,
        disguise: case.monster_type,
    };
    json!({"name": case.name, "outcome": move_monst_schedule(&mut thing, &case.results)})
}

fn chase_cases() -> Vec<ChaseCase> {
    vec![
        chase_case(
            "direct_east",
            1,
            'K',
            0,
            coord(5, 5),
            coord(5, 7),
            coord(5, 7),
        ),
        chase_case(
            "tie_break",
            7,
            'K',
            0,
            coord(5, 5),
            coord(7, 7),
            coord(9, 9),
        ),
        chase_case(
            "wall_blocks_diagonal",
            7,
            'K',
            0,
            coord(5, 5),
            coord(4, 4),
            coord(4, 4),
        )
        .tiles(vec![(coord(4, 5), '|'), (coord(5, 4), '-')]),
        chase_case(
            "scare_scroll_skip",
            7,
            'K',
            0,
            coord(5, 5),
            coord(5, 4),
            coord(5, 4),
        )
        .tiles(vec![(coord(5, 4), SCROLL)])
        .objects(vec![object(SCROLL, S_SCARE, coord(5, 4))]),
        chase_case(
            "ordinary_scroll_allowed",
            7,
            'K',
            0,
            coord(5, 5),
            coord(5, 4),
            coord(5, 4),
        )
        .tiles(vec![(coord(5, 4), SCROLL)])
        .objects(vec![object(SCROLL, 1, coord(5, 4))]),
        chase_case(
            "xeroc_skip",
            7,
            'K',
            0,
            coord(5, 5),
            coord(5, 4),
            coord(5, 4),
        )
        .monsters(vec![thing('X', coord(5, 4), 0, ':')]),
        chase_case(
            "confused_random",
            5,
            'K',
            ISHUH,
            coord(5, 5),
            coord(7, 7),
            coord(7, 7),
        ),
        chase_case(
            "phantom_random",
            1,
            'P',
            0,
            coord(5, 5),
            coord(7, 7),
            coord(7, 7),
        ),
        chase_case(
            "bat_random",
            1,
            'B',
            0,
            coord(5, 5),
            coord(7, 7),
            coord(7, 7),
        ),
    ]
}

fn move_cases() -> Vec<MoveCase> {
    vec![
        move_case("normal_one_chase", 'K', 0, true, vec![0]),
        move_case("slow_turn_skips", 'K', ISSLOW, false, vec![]),
        move_case("slow_turn_moves", 'K', ISSLOW, true, vec![0]),
        move_case("haste_two_chases", 'K', ISHASTE, true, vec![0, 0]),
        move_case("first_chase_stops", 'K', ISHASTE, true, vec![-1, 0]),
        move_case("second_chase_stops", 'K', ISHASTE, true, vec![0, -1]),
    ]
}

fn chase_case(
    name: &'static str,
    seed: i32,
    monster_type: char,
    flags: i32,
    pos: Coord,
    target: Coord,
    hero: Coord,
) -> ChaseCase {
    ChaseCase {
        name,
        seed,
        monster_type,
        flags,
        pos,
        target,
        hero,
        tiles: Vec::new(),
        objects: Vec::new(),
        monsters: Vec::new(),
    }
}

impl ChaseCase {
    fn tiles(mut self, tiles: Vec<(Coord, char)>) -> Self {
        self.tiles = tiles;
        self
    }

    fn objects(mut self, objects: Vec<ChaseObject>) -> Self {
        self.objects = objects;
        self
    }

    fn monsters(mut self, monsters: Vec<ChaseThing>) -> Self {
        self.monsters = monsters;
        self
    }
}

fn move_case(
    name: &'static str,
    monster_type: char,
    flags: i32,
    turn: bool,
    results: Vec<i32>,
) -> MoveCase {
    MoveCase {
        name,
        monster_type,
        flags,
        turn,
        results,
    }
}

impl ChaseMap {
    fn chat(&self, coord: Coord) -> char {
        self.terrain
            .iter()
            .find(|(pos, _)| coord_eq(*pos, coord))
            .map(|(_, ch)| *ch)
            .unwrap_or('.')
    }

    fn moat(&self, coord: Coord) -> Option<&ChaseThing> {
        self.monsters
            .iter()
            .find(|monster| coord_eq(monster.pos, coord))
    }

    fn winat(&self, coord: Coord) -> char {
        self.moat(coord)
            .map(|monster| monster.disguise)
            .unwrap_or_else(|| self.chat(coord))
    }

    fn object_at(&self, coord: Coord) -> Option<&ChaseObject> {
        self.objects.iter().find(|obj| coord_eq(obj.pos, coord))
    }
}

fn rndmove(rng: &mut RogueRng, thing: &ChaseThing) -> Coord {
    Coord {
        y: thing.pos.y + rng.rnd(3) - 1,
        x: thing.pos.x + rng.rnd(3) - 1,
    }
}

fn diag_ok(game_map: &ChaseMap, start: Coord, end: Coord) -> bool {
    if end.x < 0 || end.x >= NUMCOLS as i32 || end.y <= 0 || end.y >= NUMLINES as i32 - 1 {
        return false;
    }
    if end.x == start.x || end.y == start.y {
        return true;
    }
    step_ok(game_map.chat(Coord {
        y: end.y,
        x: start.x,
    })) && step_ok(game_map.chat(Coord {
        y: start.y,
        x: end.x,
    }))
}

fn dist(first: Coord, second: Coord) -> i32 {
    (second.x - first.x) * (second.x - first.x) + (second.y - first.y) * (second.y - first.y)
}

fn coord_eq(first: Coord, second: Coord) -> bool {
    first.x == second.x && first.y == second.y
}

fn push_candidate(trace: &mut Value, candidate: Value) {
    trace
        .as_object_mut()
        .unwrap()
        .get_mut("candidates")
        .unwrap()
        .as_array_mut()
        .unwrap()
        .push(candidate);
}

fn coord(y: i32, x: i32) -> Coord {
    Coord { y, x }
}

fn object(obj_type: char, which: i32, pos: Coord) -> ChaseObject {
    ChaseObject {
        obj_type,
        which,
        pos,
    }
}

fn thing(monster_type: char, pos: Coord, flags: i32, disguise: char) -> ChaseThing {
    ChaseThing {
        monster_type,
        pos,
        flags,
        turn: true,
        disguise,
    }
}

fn move_schedule_payload(thing: &ChaseThing, calls: i32, returned: i32) -> Value {
    json!({"calls": calls, "returned": returned, "thing": thing_json(thing)})
}

fn thing_json(thing: &ChaseThing) -> Value {
    json!({
        "type": char_json(thing.monster_type),
        "pos": coord_json(thing.pos),
        "flags": thing.flags,
        "turn": thing.turn,
        "disguise": char_json(thing.disguise),
    })
}

fn coord_json(coord: Coord) -> Value {
    json!({"y": coord.y, "x": coord.x})
}

fn char_json(ch: char) -> String {
    ch.to_string()
}
